src/JsonSchema/Constraints/SchemaConstraint.php

```diff
@@ -20,23 +23,70 @@
  */
 class SchemaConstraint extends Constraint
 {
+    const DEFAULT_SCHEMA_SPEC = 'http://json-schema.org/draft-04/schema#';
+
     /**
      * {@inheritdoc}
      */
     public function check(&$element, $schema = null, JsonPointer $path = null, $i = null)
     {
         if ($schema !== null) {
             // passed schema
-            $this->checkUndefined($element, $schema, $path, $i);
+            $validationSchema = $schema;
         } elseif ($this->getTypeCheck()->propertyExists($element, $this->inlineSchemaProperty)) {
             $inlineSchema = $this->getTypeCheck()->propertyGet($element, $this->inlineSchemaProperty);
             if (is_array($inlineSchema)) {
                 $inlineSchema = json_decode(json_encode($inlineSchema));
             }
             // inline schema
-            $this->checkUndefined($element, $inlineSchema, $path, $i);
+            $validationSchema = $inlineSchema;
         } else {
             throw new InvalidArgumentException('no schema found to verify against');
         }
+
+        // validate schema against whatever is defined in $validationSchema->$schema. If no
+        // schema is defined, assume self::DEFAULT_SCHEMA_SPEC (currently draft-04).
+        if ($this->factory->getConfig(self::CHECK_MODE_VALIDATE_SCHEMA)) {
+            if (!$this->getTypeCheck()->isObject($validationSchema)) {
+                throw new RuntimeException('Cannot validate non-object schema');
+            }
+            if ($this->getTypeCheck()->propertyExists($validationSchema, '$schema')) {
+                $schemaSpec = $this->getTypeCheck()->propertyGet($validationSchema, '$schema');
+            } else {
+                $schemaSpec = self::DEFAULT_SCHEMA_SPEC;
+            }
+
+            // preload standard schema specs
+            $preload = array(
+                'http://json-schema.org/draft-03/schema' => 'json-schema-draft-03.json',
+                'http://json-schema.org/draft-03/schema#' => 'json-schema-draft-03.json',
+                'http://json-schema.org/draft-04/schema' => 'json-schema-draft-04.json',
+                'http://json-schema.org/draft-04/schema#' => 'json-schema-draft-04.json'
+            );
+            $schemaStorage = $this->factory->getSchemaStorage();
+            foreach ($preload as $schemaID => $schemaFile) {
+                $schemaStorage->addSchema(
+                    $schemaID,
+                    json_decode(file_get_contents(__DIR__ . "/../../../schema-validation/$schemaFile"))
+                );
+            }
+
+            // validate schema
+            $checkMode = $this->factory->getConfig(self::CHECK_MODE_EXCEPTIONS) | self::CHECK_MODE_TYPE_CAST;
```

Comment:
To avoid complications while testing; I haven't thought through all the ramifications yet and what that would mean for possible future checkmode options that might come along. The current state of `$checkMode` here is temporary until I have considered it a bit more carefully.